File: src/935.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def __init__(self):
        self.MODULO = 10 ** 9 + 7
        self.numericCell = 10

    def knightDialer(self, n: int) -> int:
        retVal = 0

        jumps = [[4, 6], [6, 8], [7, 9], [4, 8], [3, 9, 0], [], [1, 7, 0], [2, 6], [1, 3], [2, 4]]

        previousDp = [1] * self.numericCell
        for _ in range(1, n):
            dp = [0] * self.numericCell

            for square in range(self.numericCell):
                sum = 0
                for nextSquare in jumps[square]:
                    sum = (sum + previousDp[nextSquare]) % self.MODULO

                dp[square] = sum

            previousDp = dp

        for square in range(self.numericCell):
            retVal = (retVal + previousDp[square]) % self.MODULO

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n in [1, 2, 3131]:
            # /* Example
            #  *  Input: n = 1
            #  *  Output: 10
            #  *
            #  *  Input: n = 2
            #  *  Output: 20
            #  *
            #  *  Input: n = 3131
            #  *  Output: 136006598
            #  */
            logging.info("Input: n = %s", n)

            retVal = pSolution.knightDialer(n)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)

File: src/935.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MODULO (int)(1e9 + 7)
int knightDialer(int n) {
    int retVal = 0;

    int numericCell = 10;

    struct jumpsType {
        int nextNumber[10];
        int nextNumberSize;
    } jumps[] = {{{4, 6}, 2}, {{6, 8}, 2},    {{7, 9}, 2}, {{4, 8}, 2}, {{3, 9, 0}, 3},
                 {{}, 0},     {{1, 7, 0}, 3}, {{2, 6}, 2}, {{1, 3}, 2}, {{2, 4}, 2}};

    int previousDp[numericCell];
    memset(previousDp, 0, sizeof(previousDp));
    int i;
    for (i = 0; i < numericCell; ++i) {
        previousDp[i] = 1;
    }

    int square, nextSquare, sum;
    int dp[numericCell];
    for (i = 1; i < n; ++i) {
        memset(dp, 0, sizeof(dp));

        for (square = 0; square < numericCell; ++square) {
            sum = 0;
            for (nextSquare = 0; nextSquare < jumps[square].nextNumberSize; ++nextSquare) {
                sum = (sum + previousDp[jumps[square].nextNumber[nextSquare]]) % MODULO;
            }

            dp[square] = sum;
        }

        memcpy(previousDp, dp, sizeof(previousDp));
    }

    for (square = 0; square < numericCell; ++square) {
        retVal = (retVal + previousDp[square]) % MODULO;
    }

    return retVal;
}

int main(int argc, char** argv) {
    struct testCaseType {
        int n;
    } testCase[] = {{1}, {2}, {3131}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: n = 1
     *  Output: 10
     *
     *  Input: n = 2
     *  Output: 20
     *
     *  Input: n = 3131
     *  Output: 136006598
     */

    int answer;
    int i;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: n = %d\n", testCase[i].n);

        answer = knightDialer(testCase[i].n);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}
